### PR TITLE
[training] splitting forward and backward graphs

### LINKED ISSUE #100: [training] create separate graphs in autograd phase
Currently, autograd creates backward pass in the same graph as the forward pass.

For simplicity and cleaner architecture, split the graph into two (or more - if loss and optimizer modules are included).

### PR DESCRIPTION
**Pull request by pilkicTT**
A pass is added, which takes the forge graph and creates a `ForgeGraphModule`, by splitting the forward and backward parts of the graph.

Since we are currently executing forward and backward pass as separate programs (functions) within a binary, splitting the graphs seems to be the natural approach.

Forward graph is created by cloning the forward nodes of the original graph, and adding output nodes for all the intermediate results the backward pass will need.

The backward graph is constructed from backward nodes of the original graph and utilizing the intermediate results from the forward graph. For each gradient an output node is created.

The `CompiledModel` is updated to have the state from both graphs.  Additionally, `CompiledGraphState` is cleaned up a little - removing the unused stuff.

In `lower_to_mlir()` removed the previous hacks - now we simply emit mlir function for each graph in `ForgeGraphModule` and no special hacks are needed for getting inputs/params/outputs of each graph.

This change enables MNIST Linear training to work.

Closes #100 #163

### CODE REVIEW

**Review by vladimirjovanovicTT — APPROVED** [review]

**Review by dgolubovicTT — APPROVED** [review]

File: forge/csrc/graph_lib/node_types.hpp [hunk 1]
```diff
@@ -272,17 +272,17 @@ class OutputNode : public QueueNode {
...
     {
     }
     bool requires_grad() const { return requires_grad_; }
     bool is_loss_output() const { return is_loss_output_; }
-    bool is_saved_intermediate() const { return is_saved_intermediate_; }
+    bool is_intermediate() const { return is_intermediate_; }
     bool untilize() const { return untilize_; }
     void set_requires_grad(bool requires_grad) { requires_grad_ = requires_grad; }
     void set_loss_output() { is_loss_output_ = true; }
-    void set_saved_intermediate(bool saved_intermediate) { is_saved_intermediate_ = saved_intermediate; }
```
[the remark this answers is not in the document]
Comment (pilkicTT) [reply forge/csrc/graph_lib/node_types.hpp:285]: Not sure what it previously meant, however there are currently no references to it in the repo. So, I've decided to repurpose it...

File: forge/csrc/passes/split_graph.cpp [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+        return false;
+    }
+
+    // Intermediate node is needed if an op in forward graph has a data user in the backward graph.
+    bool has_bwd_user = has_users_in_bwd(graph, node);
+
+    // Don't add duplicate intermediate output nodes.
+    auto user_edges = graph->user_data_edges(node);
+
+    bool has_output_node_already = std::any_of(
```
[the remark this answers is not in the document]
Comment (pilkicTT) [reply forge/csrc/passes/split_graph.cpp:84]: Yeah, we will create an input node from that output node in the backward graph and connect it to its users...
